plotly/src/plot.rs

```diff
@@ -417,6 +417,51 @@ impl Plot {
             .unwrap_or_else(|_| panic!("failed to export plot to {:?}", filename.as_ref()));
     }
 
+    // similar to write_image, but returns b64 string
+    #[cfg(feature = "kaleido")]
```

Comment:
@mchant `to_b64` and `to_svg` seem to be identical except the return type. Am I missing something? Can't we merge the two in a single function?

Also, the name of the function is `to_b64` but the return value is actually a `String` not an encoded base 64 value. Am I missing something?

Reply:
Hi @andrei-ng, I wasn't sure how to approach that. Although the two functions return the same type, the contents of the string are different. I was thinking that having two functions might be more intuitive for a user. to_b64 returns a string that a user can place into an img tag, for example
```
    let x: Vec<i32> = (0..100).collect();
    let y: Vec<i32> = (0..100).collect();
    
    let trace = Scatter::new(x, y);

    let mut plot = Plot::new();
    plot.add_trace(trace);

    let b64_string = plot.to_b64(ImageFormat::PNG, 300, 200, 1.0).unwrap();

    let svg_string = plot.to_svg(300, 200, 1.0);
```
b64_string's contents are
```
iVBORw0KGgoAAAANSUhEUgAAASwAAADICAYAAABS39xVAAAQ/0lEQVR4Xu2dWYxUVRqA/967qm26q1kEgXESxC2a+GBc5s2MjtHRedEQNSgalwRFQAUVktYEHR0VBTUmojGRuCSaaHxQo4mK+mBAAw8Krjgy0DaNYNMN9EZvk//UQjVt3SqLey/8nO8mHcHquuc/3//Xx7nnnnuqYnR0dFQ4IAABCBggUIGwDGSJECEAAUcAYVEIEICAGQIIy0yqCBQCEEBY1AAEIGCGAMIykyoChQAEEBY1AAEImCGAsMykikAhAAGERQ1AAAJmCCAsM6kiUAhAAGFRAxCAgBkCCMtMqggUAhBAWNQABCBghgDCMpMqAoUABBAWNQABCJghgLDMpIpAIQABhEUNQAACZgggLDOpIlAIQABhUQMQgIAZAgjLTKoIFAIQQFjUAAQgYIYAwjKTKgKFAAQQFjUAAQiYIYCwzKSKQCEAAYRFDUAAAmYIICwzqSJQCEAAYVEDEICAGQIIy0yqCBQCEEBY1AAEIGCGAMIykyoChQAEEBY1AAEImCGAsMykikAhAAGERQ1AAAJmCCAsM6kiUAhAAGFRAxCAgBkCCMtMqggUAhBAWNQABCBghgDCMpMqAoUABBAWNQABCJghgLDMpIpAIQABhEUNQAACZgggLDOpIlAIQABhUQMQgIAZAgjLTKoIFAIQQFjUAAQgYIYAwjKTKgKFAAQQFjUAAQiYIYCwzKSKQCEAAYRFDUAAAmYIICwzqSJQCEAAYVEDEICAGQIIy0yqCBQCEEBY1AAEIGCGAMIykyoChQAEEBY1AAEImCGAsMykikAhAAGERQ1AAAJmCCAsM6kiUAhAAGFRAxCAgBkCCMtMqggUAhBAWNQABCBghgDCMpMqAoUABBAWNQABCJghgLDMpIpAIQABhEUNQAACZgggLDOpIlAIQABhUQMQgIAZAgjLTKoIFAIQQFjUAAQgYIYAwjKTKgKFAAQQFjUAAQiYIYCwzKSKQCEAAYRFDUAAAmYIICwzqSJQCEAAYVEDEICAGQIIy0yqCBQCEEBY1AAEIGCGAMIykyoChQAEEBY1AAEImCGAsMykikAhAAGERQ1AAAJmCCAsM6kiUAhAAGFRAxCAgBkCCMtMqggUAhBAWNQABCBghgDCMpMqAoUABBAWNQABCJghgLDMpIpAIQABhEUNQAACZgggLDOpIlAIQABhUQMQgIAZAgjLTKoIFAIQQFjUAAQgYIYAwjKTKgKFAAQQFjUAAQiYIYCwzKSKQCEAAe+E9f3338uyZctk5cqVMmvWrFwFfPfdd7Jq1So5cOCATJ8+Xe6//36ZOHGiez3oNUoIAhCIj4BXwvr666/liSeekKqqKmltbc0Ja2RkRObNmyf33nuvnH322fL+++/L+vXrZcWKFRL0WnxpoiUIQEAJeCWsp59+Wq644gp57rnn5I477sgJS0dQa9askdWrV7uqGB0dlTlz5sjLL78s27dvL/haQ0MDVQQBCMRIwCthZbkuXrxY7rzzzpywPv74Y9m0aZMsXbo0h15fX7BggbS1tRV87bTTTosxVTQFAQggLBF3Cbh161ZZuHBhriKWLFkic+fOlfb29oKvffPNN240dsMNN8RSSUNDQ9LX1yeNjY2xtKeNdHV1SVNTk1RUVMTSZk9Pj9TU1EhtbW0s7R08eFAGBwclrNFyz8Fh2f57r7Tt7ZUd+tOpP33uz129g7Jh+d9j6dfx2gjCEpFPPvlENm7cOGaENX/+fFm0aJETVqHXvvrqK4QV8ifjWBfWwNCItHelBeREpELKialX9vYOBhLZ9p9/hkzMr9MhLBH56aef3LxWdg5LJ9qvvvpqWbt2rXR0dBR87Z133kFYIX9ejrawhkdGZWd3/7jRUXqk1Cu/7R8I7HFddaXMTCVlZktCZrYkZYb7c1JmphLyl5akTEjUhEzMr9MhLBF3J/DWW28VndvK3iVct26du6MY9Norr7yCsEL+vMQhrN0HBnKjo2179sv/9hyQjv2DsmNvn+zs6pOhkdGCvaqurJBpzQknICciJ6PMf1sSMqmhTmK6eg6ZvI3TIaxMnn7++WcnqH379smUKVPcOqypU6e6Vwu9hrDCL/IwhLWvLy0fvVRry8wf5V+69Q+OBAY+pbHuMBll5JRKyrSmeqmqjGc+L3y69s/opbD+KG26Bmvbtm1SWVnpXr788stzk+mFFo4irPA/AKUIS4XjJrVz80dj55L29Q8FBtaUqMldpp3UVCf6M+vE5swlXEL0so7j2CSAsDJ5ue2229wIS++I5R9BC0cRVvhFrcKqqKqW33uH06OkzNxR/h03vaQLOpK1VXmXaodGRzqfNCOVkMb66tzbw75LGD4RzphPAGFlaFx77bXy+uuvj7t9H7So9O2332YOq4zP0+ioSHoeafwoaXtnj3TsGxCd/C501FRVyPTmxKEJ7cykdnZOaWJD6UsiEFYZCTyKb0FYGfjXXXede3ZQ57AmT57sFpaefPLJErSo9Msvv0RYBYpXb+8fPjpqy4yY9L8HhwvPI+mk9YmN9Zl5JJ3gzk5qpye4pzbVS1jTSAjrKNqnjKYR1h9A++ijj+TVV191j+YELSrVhaNa8Do6i+PQy1NdPBrXokrt08DAgGvv8IWjfYPD8mtXv7Tt7ZdfdRlAZ2/671197v/pAsqgoyVZ4y7PpjfXywz9SemIKSEnNlTLtOZ6SdTGc/t/eHjY3QnWxapxHMlkMo5mjts2EFaB1F5zzTXyzDPPyObNmwMXjqpAdHQWx6EfLF2VXVdXF3lzOgLq6B6Q//7WLb/1DDshtTlBqZD6pLPIAskT6qplRkpllJaSW5PkLuPSckrUVP1hH/QfAH04XX/iOFRY+hPXPwJx5C4ObkerDYRVgLwuHNUR1s6dO4/LhaM6RdQxZoFkdsV2eqJ71/5+0bmmQkdtVaUTT3px5KGJ7ew8UipZ3oillLuEYX5YuCQMk2b050JYItLd3e0ufXT9lR66gv2LL76Qxx9/3PTC0d97DubNIx2646ZLAn7t6pPB4cJG0rVGuuZoWmOt/HVyo/xlYv4CyaRMPiGaBZIIK/oPveUWEJaIe/zm4Ycfdpv36Tos3djv9ttvl1Qq5XJ7rC4c3d8/lLlEG/+grY6SeovMI6l0so+QjJ3YTrjV3Lqqm4efLX+8j7/YEVYJOT1aC0f1QdvsnTVdh6SPkGzbvV86Dgy5kVN3X/CDthPqq8c9OpIVk65Jqq8pvkASYZVQIPxKbAQQVhHUUS4cPdIHbVU4Y55l04WR+tCtLgNIJUJ50BZhxfZZpKESCCCsIpCOZOGoTlrv6Tn0oG3+NiTlPGird9v0tv/sk1rcpZxe0kV9IKyoCXP+P0MAYRWhVWzhaN+QyN/+8a/cYyS5Z9wyD93qZV3Q8WcetGUDvz9T2qX9LncJS+N0rPwWwiqSiWILR1u3tASeIVk1IqnaEWmuGZZUzYi01I1Kqjb956aaEamuCFg7cNiZj8bC0f7+fqmvr4+tXnWdmd74iHMdVpwLR6+//vrYWB6PDSGsIlkN2o1Udxz99+YGmVBf5QTkRFQ74v7cnPl7XWXpQipWYHEuHM3GosLSxY5xbZF8vC8cjWs77WK1ZPV1hFUkc0G7kca5t7rVAiNuCIRJAGEVoRm042iYieBcEIBAcQIIqzijggtHS3grvwIBCIRIAGGFCJNTQQAC0RJAWGXyLbT6vczTjXmb3il744035PPPP3fPMk6aNEnuvvvu3LOOUbatgXR2dsott9zi9gS76KKLXGxRtfnWW2/Jhx9+6Nq48MIL5aabboqsPb1Jog+063dJVldXu2//zn4ZblT9C6MeOMchAgirjGoIWv1exunGvaW3t9d9V+Kll17q9mnSD/W3334rra2tTmDz5s0T3YM++w0/69evlxUrVoTRtDvHAw88IHp38LLLLnPCiqrNN998011u33XXXWOWTkTRnm4ho1sG6de56UPuupeZbh/04osvRta/0BLCiXIEEFYZxRC0+j2sbxDOD0s/1I899pi88MILbqSzZs2a3Hco6mhhzpw5buQQRts62vnhhx/cCOSMM85wwoqiTRXIjTfeKM8///y4uKNoT3eB0NGbjlx1iYbuLKtf7aZ/j6K9MsqKt5RAAGGVAOnwXwla/Z69xCjjtAXfotvdbN26VZYsWRK4ZfORtr179243ulq1apUbeZx11llOWFH095dffpGVK1fKKaec4kZZuopfhXL++edH0p7CffTRR0V3/FTBqyjPPfdcufLKKyNrL8wa4FxpAgirjEoIWv1+zjnnlHHGwm9Ridxzzz1uby79nsQo2162bJnbPVUvNZ999tmcsKJoU+eT9DJWv6no9NNPl127dsmiRYvkpZdeks8++8wJeuHChTkwKuu5c+fKkfDds2ePLFiwwF3u6ujxwQcfdJeiUfQv1CLgZFwSHkkNBK1+1w9fWIdetixdulRuvvlmOe+889xpo2r73XfflR07dsj8+fNdO/nCiqLNTZs2yWuvvSZPPvlkDpfOy+n8nMpr48aNru/ZQ+NSoZXLV/c60/erBM8880w3L6ibNK5evVo+/fTT0NsLqwY4z1gCjLDKqIg4Vr/rnMt9990nV111Ve5OnYYaVdsPPfSQbNmyJUcju/PnJZdcIhdffHHBbaLLXe3f3t4uOqJbu3Ztrs2sUHTSXSfHVSZ66N91y2r93XLbUylt2LDBMc0eOtrSCf8o2iujrHhLCQQQVgmQDv+VqFe/6yXL8uXL3bdPqyzyj6jbzraVP8KKqk2VhfZRpfjjjz+6XV/1klAffNYJ8cWLF+fuhK5bt85dPpZ76MT6U0895e4MJhIJ0ctDHW3pDQy9WRF2e+XGyfuCCSCsMiuk0LbJZZ5uzNvee+8998GaMGHCmP//yCOPyOzZs2NZeZ8vLA0iiv7qKEslpHvqqzR03depp57q+hxFe/rFtx988EHuK71054QLLrggsvbCqAXOwSUhNQABCBglwAjLaOIIGwI+EkBYPmadPkPAKAGEZTRxhA0BHwkgLB+zTp8hYJQAwjKaOMKGgI8EEJaPWafPEDBKAGEZTRxhQ8BHAgjLx6zTZwgYJYCwjCaOsCHgIwGE5WPW6TMEjBJAWEYTR9gQ8JEAwvIx6/QZAkYJICyjiSNsCPhIAGH5mHX6DAGjBBCW0cQRNgR8JICwfMw6fYaAUQIIy2jiCBsCPhJAWD5mnT5DwCgBhGU0cYQNAR8JICwfs06fIWCUAMIymjjChoCPBBCWj1mnzxAwSgBhGU0cYUPARwIIy8es02cIGCWAsIwmjrAh4CMBhOVj1ukzBIwSQFhGE0fYEPCRAMLyMev0GQJGCSAso4kjbAj4SABh+Zh1+gwBowQQltHEETYEfCSAsHzMOn2GgFECCMto4ggbAj4SQFg+Zp0+Q8AoAYRlNHGEDQEfCSAsH7NOnyFglADCMpo4woaAjwQQlo9Zp88QMEoAYRlNHGFDwEcCCMvHrNNnCBglgLCMJo6wIeAjAYTlY9bpMwSMEkBYRhNH2BDwkQDC8jHr9BkCRgkgLKOJI2wI+EgAYfmYdfoMAaMEEJbRxBE2BHwkgLB8zDp9hoBRAgjLaOIIGwI+EkBYPmadPkPAKAGEZTRxhA0BHwkgLB+zTp8hYJQAwjKaOMKGgI8EEJaPWafPEDBKAGEZTRxhQ8BHAgjLx6zTZwgYJYCwjCaOsCHgIwGE5WPW6TMEjBJAWEYTR9gQ8JEAwvIx6/QZAkYJICyjiSNsCPhIAGH5mHX6DAGjBBCW0cQRNgR8JICwfMw6fYaAUQIIy2jiCBsCPhJAWD5mnT5DwCiB/wPNgWdOqPm8EwAAAABJRU5ErkJggg==
```
which can be used in a tag like this
```
<img src="data:image/png;base64,{b64_string}" />
```
svg_string's contents are
```
<svg class=\"main-svg\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"300\" height=\"200\" style=\"\" viewBox=\"0 0 300 200\"><rect x=\"0\" y=\"0\" width=\"300\" height=\"200\" style=\"fill: rgb(255, 255, 255); fill-opacity: 1;\"/><defs id=\"defs-f5e43f\"><g class=\"clips\"><clipPath id=\"clipf5e43fxyplot\" class=\"plotclip\"><rect width=\"140\" height=\"20\"/></clipPath><clipPath class=\"axesclip\" id=\"clipf5e43fx\"><rect x=\"80\" y=\"0\" width=\"140\" height=\"200\"/></clipPath><clipPath class=\"axesclip\" id=\"clipf5e43fy\"><rect x=\"0\" y=\"100\" width=\"300\" height=\"20\"/></clipPath><clipPath class=\"axesclip\" id=\"clipf5e43fxy\"><rect x=\"80\" y=\"100\" width=\"140\" height=\"20\"/></clipPath></g><g class=\"gradients\"/></defs><g class=\"bglayer\"/><g class=\"layer-below\"><g class=\"imagelayer\"/><g class=\"shapelayer\"/></g><g class=\"cartesianlayer\"><g class=\"subplot xy\"><g class=\"layer-subplot\"><g class=\"shapelayer\"/><g class=\"imagelayer\"/></g><g class=\"gridlayer\"><g class=\"x\"><path class=\"xgrid crisp\" transform=\"translate(108.28,0)\" d=\"M0,100v20\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xgrid crisp\" transform=\"translate(136.57,0)\" d=\"M0,100v20\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xgrid crisp\" transform=\"translate(164.85,0)\" d=\"M0,100v20\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xgrid crisp\" transform=\"translate(193.13,0)\" d=\"M0,100v20\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/></g><g class=\"y\"><path class=\"ygrid crisp\" transform=\"translate(0,109.91)\" d=\"M80,0h140\" style=\"stroke: rgb(238, 238, 238); stroke-opacity: 1; stroke-width: 1px;\"/></g></g><g class=\"zerolinelayer\"><path class=\"xzl zl crisp\" transform=\"translate(80,0)\" d=\"M0,100v20\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"yzl zl crisp\" transform=\"translate(0,119)\" d=\"M80,0h140\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/></g><path class=\"xlines-below\"/><path class=\"ylines-below\"/><g class=\"overlines-below\"/><g class=\"xaxislayer-below\"/><g class=\"yaxislayer-below\"/><g class=\"overaxes-below\"/><g class=\"plot\" transform=\"translate(80,100)\" clip-path=\"url('#clipf5e43fxyplot')\"><g class=\"scatterlayer mlayer\"><g class=\"trace scatter trace0be7b6\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,19L140,1\" style=\"vector-effect: non-scaling-stroke; fill: none; stroke: rgb(31, 119, 180); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g></g></g><g class=\"overplot\"/><path class=\"xlines-above crisp\" d=\"M0,0\" style=\"fill: none;\"/><path class=\"ylines-above crisp\" d=\"M0,0\" style=\"fill: none;\"/><g class=\"overlines-above\"/><g class=\"xaxislayer-above\"><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"133\" transform=\"translate(80,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">0</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"133\" transform=\"translate(108.28,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">20</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"133\" transform=\"translate(136.57,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">40</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"133\" transform=\"translate(164.85,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">60</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"133\" transform=\"translate(193.13,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">80</text></g></g><g class=\"yaxislayer-above\"><g class=\"ytick\"><text text-anchor=\"end\" x=\"79\" y=\"4.199999999999999\" transform=\"translate(0,119)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">0</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"79\" y=\"4.199999999999999\" transform=\"translate(0,109.91)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">50</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"79\" y=\"4.199999999999999\" transform=\"translate(0,100.82)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(68, 68, 68); fill-opacity: 1; white-space: pre;\">100</text></g></g><g class=\"overaxes-above\"/></g></g><g class=\"polarlayer\"/><g class=\"ternarylayer\"/><g class=\"geolayer\"/><g class=\"funnelarealayer\"/><g class=\"pielayer\"/><g class=\"treemaplayer\"/><g class=\"sunburstlayer\"/><g class=\"glimages\"/><defs id=\"topdefs-f5e43f\"><g class=\"clips\"/></defs><g class=\"layer-above\"><g class=\"imagelayer\"/><g class=\"shapelayer\"/></g><g class=\"infolayer\"><g class=\"g-gtitle\"/><g class=\"g-xtitle\"/><g class=\"g-ytitle\"/></g></svg>
```
the contents svg_string has to be rendered as an svg object itself.

Reply:
Got it! All clear now. I will refactor the functions into one, since they are both very similar and add will merge it after that.

Reply:
Looking forward to this.

Up until now, I've been writing to file and then loading it again.

It will be convenient to finally be able to get the raw svg string.